index.bs

```diff
@@ -4514,9 +4514,9 @@ In order to perform a [=registration ceremony=], the [=[RP]=] MUST proceed as fo
 
 1. Check that the <code>[=credentialId=]</code> is &le; 1023 bytes. Credential IDs larger than this many bytes SHOULD cause the RP to fail this [=registration ceremony=].
 
-1. Check that the <code>[=credentialId=]</code> is not yet registered to any other user. If registration
-    is requested for a credential that is already registered to a different user, the [=[RP]=] SHOULD
-    fail this [=registration ceremony=], or it MAY decide to accept the registration, e.g. while deleting the older registration.
+1. Check that the <code>[=credentialId=]</code> is not yet registered for any user. If it is, the [=[RP]=] MUST fail this [=registration ceremony=].
+
+    NOTE: [=credential IDs=] contain sufficient entropy that the case of accidential duplication can be ignored. However, several [=attestation=] formats do not include a self-signature and thus an attacker, with access to a site's credential database, could attempt to register a victim's credential as their own. If the [=[RP]=] accepted this and replaced the existing registration, and [=discoverable credentials=] were used, then the victim could be forced to sign into the attackers account at their next attempt. Data saved by the victim in that state would then be available to the attacker. Thus duplicate credential IDs are rejected in this step.
```

Comment:
It took me a few readings to understand that the first sentence wasn't negating the "MUST". What about something like that?
```suggestion
    NOTE: [=credential IDs=] contain sufficient entropy that an accidental ID collision is unlikely. However, several [=attestation=] formats do not include a self-signature and thus an attacker, with access to a site's credential database, could attempt to register a victim's credential as their own. If the [=[RP]=] accepted this and replaced the existing registration, and [=discoverable credentials=] were used, then the victim could be forced to sign into the attackers account at their next attempt. Data saved by the victim in that state would then be available to the attacker. Thus duplicate credential IDs are rejected in this step.
```
(Also, is this completely true, given randomness issues with low-power devices / virtual environments / OS bugs?)

Reply:
I prefer @lgarron's suggested first sentence re-write.

Reply:
Some modest suggested clarifications:
1. "several [=attestation=] formats" --- I think what you mean here is "[attestation types](https://www.w3.org/TR/webauthn/#sctn-attestation-types)".  IIUC, all attestation types other than "self attestation" do not include a self-signature by the cred private key.
2. an attacker would not necessarily need to have "access to a site's credential database", rather they would just need to have somehow obtained the victim's site-specific credential ID(s), which could e.g. also be done via infecting the user's browser in some fashion.
3. I'd state up front that this is rationale for RPs to simply reject duplicate cred IDs.
4. I would not want this potential attack exposition to scare off RPs from using WebAuthn.   Am not sure whether or how to additionally say something along the lines of "obtaining user cred IDs, eg by site account dbase compromise or browser compromise, does not give an attacker _direct_ access to victims' accounts, unlike the case of obtaining users' username+passwords."

This likely needs further polishing (and @lgarron's first sentence fix from above):
```suggestion
    NOTE: The rationale for [=[RPS]=] rejecting duplicate [=credential IDs=] is as follows: [=credential IDs=] contain sufficient entropy that the case of accidental duplication can be ignored. However, [=attestation types=] other than [=self attestation=] do not include a self-signature explicitly proving possession of the [=credential private key=] at [=registration=] time. Thus an attacker who has managed to obtain a user's [=credential ID=] for a site (this could be potentially accomplished in various ways), could attempt to register a victim's credential as their own at that site. If the [=[RP]=] accepts this new registration and replaces the victim's existing credential registration, and the [=discoverable credentials|credentials are discoverable=], then the victim could be forced to sign into the attacker's account at their next attempt. Data saved to the site by the victim in that state would then be available to the attacker.
```

Reply:
I'd add to @equalsJeffH's suggestion that the attacker would also need the victim's public key:

```
[...] Thus an attacker who has managed to obtain a user's [=credential ID=] and [=credential public key=] for a site  [...]
```

Reply:
Have applied Jeff's suggestion and updated it with Emil's note that the public key is also needed.

I'm not sure if Lucas was proposing an explicit change for the first line but I've rewritten it to avoid the negative potentially applying to the MUST. (I also changed it to a SHOULD because we're giving the reasoning and thus kinda suggesting to the RP why it might be optional.)